rafs/src/builder/core/node.rs

```diff
@@ -387,32 +391,77 @@ impl Node {
         chunk.set_uncompressed_offset(pre_uncompressed_offset);
         chunk.set_uncompressed_size(uncompressed_size);
 
-        let compressed_size = if ctx.blob_features.contains(BlobFeatures::SEPARATE) {
-            // For `tar-ref`, `targz-ref`, `estargz-ref` and `estargzindex-ref`.
-            chunk.compressed_size()
-        } else {
-            // For other case which needs to write chunk data to data blobs.
-            let (compressed, is_compressed) = compress::compress(chunk_data, ctx.compressor)
-                .with_context(|| format!("failed to compress node file {:?}", self.path()))?;
-            let compressed_size = compressed.len() as u32;
-            let pre_compressed_offset = blob_ctx.current_compressed_offset;
-            blob_writer
-                .write_all(&compressed)
-                .context("failed to write blob")?;
-            blob_ctx.blob_hash.update(&compressed);
-            blob_ctx.current_compressed_offset += compressed_size as u64;
-            blob_ctx.compressed_blob_size += compressed_size as u64;
+        let mut chunk_info = None;
+
+        if self.inode.child_count() == 1
+            && uncompressed_size < ctx.batch_size / 2
+            && ctx.blob_batch_generator.is_some()
+        {
+            // This chunk will be added into a batch chunk.
+            let mut batch = ctx.blob_batch_generator.as_ref().unwrap().lock().unwrap();
+
+            if batch.chunk_data_buf_len() as u32 + uncompressed_size < ctx.batch_size {
+                // Add into current batch chunk directly.
+                batch.append_chunk_data_buf(chunk_data.to_vec().as_mut());
```

Comment:
Function  `append_chunk_data_buf()` should be refined to take a `&[u8]` instead of a `Vec[u8]` to avoid unnecessary memory copy.

Reply:
Refined.